test/integration/record_linkage.py

```diff
@@ -0,0 +1,211 @@
+        scale=sigma, size=sum(noisy_rows)
+    )
+    return col + modifier_array
+
+
+def _modify_categorical(
+    df: pd.DataFrame, col: str, categories: list, probability: float = 0.01
+) -> pd.DataFrame:
+    """Randomly modify categorical column using given probability."""
+    modify_elements = _RANDOM_GENERATOR.random(len(df)) > (1 - probability)
+    df.loc[modify_elements, col] = _RANDOM_GENERATOR.choice(categories)
+    return df
+
+
+def _generate_random_test_df(
+    default_plant_name: str,
+    size: int = 2022 - 1994,
+    plant_name_max_edits: int = 5,
+    plant_type_error_prob: float = 0.01,
+    construction_type_error_prob: float = 0.01,
+    construction_year_error_prob: float = 0.01,
+    capacity_mean: float = 500.0,
+    capacity_sigma: float = 10.0,
+    capacity_change_prob: float = 0.01,
+    utility_id_error_prob: float = 0.01,
+    utility_id: int = random.randrange(1000),
+):
+    """Generate a random input DataFrame for testing record linkage."""
+
+    generated_df = pd.DataFrame(
+        {
+            "base_plant_name": [default_plant_name] * size,
+            "plant_type": [
+                random.choice(list(PLANT_TYPE_CATEGORIES["categories"].keys()))
+            ]
+            * size,
+            "report_year": list(range(1994, 1994 + size)),
+            "construction_type": [
+                random.choice(list(CONSTRUCTION_TYPE_CATEGORIES["categories"].keys()))
+            ]
+            * size,
+            "capacity_mw": [capacity_mean] * size,
+            "construction_year": [
+                random.randrange(
+                    VALID_PLANT_YEARS["lower_bound"], VALID_PLANT_YEARS["upper_bound"]
+                )
+            ]
+            * size,
+            "utility_id_ferc1": [utility_id] * size,
+        }
+    )
+
+    # Add random edits to plant name
+    generated_df["plant_name_ferc1"] = generated_df["base_plant_name"].apply(
+        _randomly_modify_string
+    )
+
+    # Modify capacity rows based on probability and sigma
+    generated_df["capacity_mw"] = _noisify(
+        generated_df["capacity_mw"],
+        sigma=capacity_sigma,
+        probability=capacity_change_prob,
+    )
+
+    # Modify categorical columns
+    generated_df = _modify_categorical(
+        generated_df,
+        "construction_type",
+        list(CONSTRUCTION_TYPE_CATEGORIES["categories"]),
+        construction_type_error_prob,
+    )
+    generated_df = _modify_categorical(
+        generated_df,
+        "construction_year",
+        list(range(VALID_PLANT_YEARS["lower_bound"], VALID_PLANT_YEARS["upper_bound"])),
+        construction_year_error_prob,
+    )
+    generated_df = _modify_categorical(
+        generated_df,
+        "plant_type",
+        list(PLANT_TYPE_CATEGORIES["categories"]),
+        plant_type_error_prob,
+    )
+
+    # Generate l1 unit vector to represent fuel fractions (all add up to 1)
+    fuel_frac_vec = abs(_RANDOM_GENERATOR.normal(size=len(_FUEL_COLS)))
```

Comment:
Also probably not a great way to model the fuel fractions where plants tend to be heavily weighted towards one fuel, not distributed radomly across all types